File: src/dotnet/PolyBus/PolyBus.cs
using PolyBus.Transport;
using PolyBus.Transport.Transactions;
using PolyBus.Transport.Transactions.Messages;
using PolyBus.Transport.Transactions.Messages.Handlers;

namespace PolyBus;

public class PolyBus(PolyBusBuilder builder) : IPolyBus
{
    public IDictionary<string, object> Properties => builder.Properties;

    public ITransport Transport { get; set; } = null!;

    public IList<IncomingHandler> IncomingPipeline { get; } = builder.IncomingPipeline;

    public IList<OutgoingHandler> OutgoingPipeline { get; } = builder.OutgoingPipeline;

    public Messages Messages { get; } = builder.Messages;

    public Task<IncomingTransaction> CreateIncomingTransaction(IncomingMessage message) =>
        builder.IncomingTransactionFactory(builder, this, message);

    public  Task<OutgoingTransaction> CreateOutgoingTransaction() =>
        builder.OutgoingTransactionFactory(builder, this);

    public async Task Send(Transaction transaction)
    {
        var step = () => Transport.Handle(transaction);

        if (transaction is IncomingTransaction incomingTransaction)
        {
            var handlers = transaction.Bus.IncomingPipeline;
            for (var index = handlers.Count - 1; index >= 0; index--)
            {
                var handler = handlers[index];
                var next = step;
                step = () => handler(incomingTransaction, next);
            }
        }
        else if (transaction is OutgoingTransaction outgoingTransaction)
        {
            var handlers = transaction.Bus.OutgoingPipeline;
            for (var index = handlers.Count - 1; index >= 0; index--)
            {
                var handler = handlers[index];
                var next = step;
                step = () => handler(outgoingTransaction, next);
            }
        }

        try
        {
            await step();
        }
        catch
        {
            await transaction.Abort();
            throw;
        }
    }

    public Task Start() => Transport.Start();

    public Task Stop() => Transport.Stop();

    public string Name => builder.Name;
}

File: src/kotlin/src/main/kotlin/polybus/PolyBus.kt
package polybus

import polybus.transport.ITransport
import polybus.transport.transactions.IncomingTransaction
import polybus.transport.transactions.OutgoingTransaction
import polybus.transport.transactions.Transaction
import polybus.transport.transactions.messages.IncomingMessage
import polybus.transport.transactions.messages.Messages
import polybus.transport.transactions.messages.handlers.IncomingHandler
import polybus.transport.transactions.messages.handlers.OutgoingHandler

class PolyBus(private val _builder: PolyBusBuilder) : IPolyBus {
    override val properties: MutableMap<String, Any>
        get() = _builder.properties

    override lateinit var transport: ITransport

    override val incomingPipeline: MutableList<IncomingHandler> = _builder.incomingPipeline
    override val outgoingPipeline: MutableList<OutgoingHandler> = _builder.outgoingPipeline
    override val messages: Messages = _builder.messages

    override suspend fun createIncomingTransaction(message: IncomingMessage): IncomingTransaction =
        _builder.incomingTransactionFactory(_builder, this, message)

    override suspend fun createOutgoingTransaction(): OutgoingTransaction =
        _builder.outgoingTransactionFactory(_builder, this)

    override suspend fun send(transaction: Transaction) {
        var step: suspend () -> Unit = { transport.handle(transaction) }

        when (transaction) {
            is IncomingTransaction -> {
                for (index in transaction.bus.incomingPipeline.indices.reversed()) {
                    val handler = transaction.bus.incomingPipeline[index]
                    val next = step
                    step = { handler(transaction, next) }
                }
            }

            is OutgoingTransaction -> {
                for (index in transaction.bus.outgoingPipeline.indices.reversed()) {
                    val handler = transaction.bus.outgoingPipeline[index]
                    val next = step
                    step = { handler(transaction, next) }
                }
            }
        }

        try {
            step()
        } catch (error: Throwable) {
            transaction.abort()
            throw error
        }
    }

    override suspend fun start() {
        transport.start()
    }

    override suspend fun stop() {
        transport.stop()
    }

    override val name: String
        get() = _builder.name
}
